update url redaction logic for cloud vs dotcom

akalia25 opened the pull request:
This PR updates the URL redaction logic for cookie values to handle cloud vs dotcom appropriately:
1. Removes URL redaction for dotcom events so marketing fields like `first_source_url` are stored unmodified. This is safe now that dotcom does not host private code. 
2. For private instances, the marketing cookie vars will be blank 
3. Migrates redaction logic from front-end to back-end
## Test plan
- Updated tests in `event_handlers`.

- Verified in local `event_logs` table that marketing fields are not being collected for private instances by running `sg start enterprise` and testing against `https://sourcegraph.test:3443`.

- The `url` field is not redacted as it is required.

This ensures we maintain marketing attribution tracking on dotcom while preventing the storing of sensitive URLs from cloud instances.

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles 

Why does it matter? 

These test plans are there to demonstrate that are following industry standards which are important or critical for our customers. 
They might be read by customers or an auditor. There are meant be simple and easy to read. Simply explain what you did to ensure 
your changes are correct!

Here are a non exhaustive list of test plan examples to help you:

- Making changes on a given feature or component: 
  - "Covered by existing tests" or "CI" for the shortest possible plan if there is zero ambiguity
  - "Added new tests" 
  - "Manually tested" (if non trivial, share some output, logs, or screenshot)
- Updating docs: 
  - "previewed locally" 
  - share a screenshot if you want to be thorough
- Updating deps, that would typically fail immediately in CI if incorrect
  - "CI" 
  - "locally tested" 
-->

dadlerj commented:
I haven't reviewed the actual changes, but I will say that this same logical change should broadly apply to all marketing variables, right? Not just the two session vars you listed above? I see several other places in this file that we do similar hostname-based checks and redactions. Curious if there's any reason NOT to do it there as well (and for other, non-marketing related props such as the page URL, a top level field in the event object).

nathan-downs commented:
I agree with Dan's comment re:other marketing fields, we're not collecting it on our side so we shouldn't be logging it.

nathan-downs commented:
When you run with `sg start` and create a test instance does it redact correctly?

akalia25 commented:
> I haven't reviewed the actual changes, but I will say that this same logical change should broadly apply to all marketing variables, right? Not just the two session vars you listed above? I see several other places in this file that we do similar hostname-based checks and redactions. Curious if there's any reason NOT to do it there as well (and for other, non-marketing related props such as the page URL, a top level field in the event object).

The same logic should apply! And I could update the scope of the PR to apply it to our other fields (`first_source_url`, `last_source_url`, etc). 

From my perspective, I kept the scope of this PR limited to ensure that these changes are stable first before expanding further.

akalia25 commented on client/web/src/tracking/eventLogger.ts:
```diff
@@ -27,6 +27,8 @@ export const SESSION_FIRST_URL_KEY = 'sessionFirstUrl'
 
 const EXTENSION_MARKER_ID = '#sourcegraph-app-background'
 
+const isSourcegraphWebSiteId = window.context?.siteID === 'SourcegraphWeb'
+
```
Use this to verify if its a customer's instance or a dotcom webpage.

akalia25 commented on client/web/src/tracking/eventLogger.ts:
```diff
@@ -349,18 +325,7 @@ export class EventLogger implements TelemetryService, SharedEventLogger {
-            return referrer
-        } catch {
-            return ''
-        }
+        return referrer
```
we don't need to redact `referrer` on dotcom anymore since no private code is hosted there. Additionally, redaction for this field is done on `event_handlers.go` before sending to BQ

akalia25 commented on client/web/src/tracking/eventLogger.ts:
```diff
@@ -258,59 +262,38 @@ export class EventLogger implements TelemetryService, SharedEventLogger {
-            return this.originalReferrer
-        }
+
+        cookies.set(ORIGINAL_REFERRER_KEY, originalReferrer, this.cookieSettings)
```
redaction for original referrer done on `event_handlers` side

akalia25 replied:
Great question! I redacted the `referrer` in the backend instead of the frontend for two reasons:

1. The existing referrer redaction in the frontend used custom regex logic rather than the `redactSensitiveInfoFromAppURL` function. I thought it made more sense to consolidate this logic in the backend.
2. I treated `referrer` like the `URL` field - redacting it in the backend and sending the unredacted value to the cloud database. This way, if a developer checks the cookie values on a cloud instance, the frontend cookie matches what they see in Postgres.

I kept the existing redaction logic for `first_source_url`, `last_source_url`, and `session_first_url` in the frontend, since those already used `redactSensitiveInfoFromAppURL`. My goal was to minimize changes and move outdated custom redaction logic to the backend, while maintaining existing patterns.

Talking it through, i'll update it to have all the redaction be done in the back-end and send redacted versions to BQ :)

dadlerj reviewed (changes requested):
Thank you for this! A few small comments, and a few big ones. I want to propose a bit of restructuring though, since I'm afraid that this setup will be a bit brittle (seeing us check if URLs contain specific substrings like "github.com/..." and ".sourcegraph." worries me).

How does the following sound to simplify it:

* In the frontend:
   * If dotcom, send `url` and all marketing URL vars to the backend unredacted.
   * If cloud, send `url` unredacted, but don't send ANY marketing URL vars (unless you can think of a reason we want them?)
   * In essence: zero redaction occurs on the frontend and we can centralize it all in the backend
* In the backend:
   * If dotcom, no redaction on the `url` var or any of the marketing vars
   * If cloud, there are no marketing vars. For `url`, we redact using roughly the logic here: https://github.com/sourcegraph/sourcegraph/pull/57384/files#diff-442914ea375af7f025d973a5aec063bcf9814f198b0f057eee99efed7f5862c6R47-R59. The one thing I want to ask is that we expand the list of potential URLs that we pattern match against. The problem is that MANY of our customers don't have `github.com/...` or `gitlab.com/...` prefixes to their code host URLs. They likely host their code internally on an internal domain (e.g. `github.uber.com/...` or just `code.uber.com/...`) so there's no easy way to know what to redact. Let's, at the very least, capture all of Sourcegraph's built-in URL patterns (e.g. notebooks and batch changes and insights and admin pages and user pages and settings pages and etc.) so if we just see the URL as /redacted we know it's most likely the case that it was a custom code view page. Or perhaps instead of trying to list out every possible path that doesn't need to be redacted, we can just always take the first part of the path and redact anything after it—e.g. if the url is `uber.sourcegraphcloud.com/code.uber.com/my-repo-name`, we would take the hostname (uber.sourcegraphcloud.com/), the _first_ part of the path (code.uber.com/), but redact everything after that... Seems like that would work with code views, searches, sign-in, etc.? What do you think?

akalia25 commented on client/web/src/tracking/eventLogger.ts:
```diff
@@ -27,6 +27,8 @@ export const SESSION_FIRST_URL_KEY = 'sessionFirstUrl'
 
 const EXTENSION_MARKER_ID = '#sourcegraph-app-background'
 
+const isSourcegraphWebSiteId = window.context?.siteID === 'SourcegraphWeb'
+
```
Initially I was using the `sourcegraphDotComMode` ([commit ref](https://github.com/sourcegraph/sourcegraph/pull/57384/commits/7c51453ca3cbdbe0c03211c3f284c95cd4a9c979)) but upon testing I discovered that `sourcegraphDotComMode` does not distinguish between a cloud instance and dotcom. Below is a screenshot of what happens when I use the `window.context?.sourcegraphDotComMode` reference, the redaction fails on the cloud instance

<img width="1396" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/32119652/64855986-c496-4fe9-8ecb-3fcaf3e1405f">

dadlerj replied:
WOAH that is shocking to me and seems like a big issue. Let me raise that to the cloud team to see if it's by design...

dadlerj replied:
Just to confirm—how are you _testing_ what would happen in a cloud instance?

akalia25 replied:
I tested this by creating a local SG instance (`sg start dotcom`) and querying the local `event_logs` tables. The URL for the test instance is `https://sourcegraph.test:3443`.

Sorry I may have over-generalized to say it fails on cloud instances, but it fails on local instances. From my understanding `dotcomMode` should not be true for `https://sourcegraph.test:3443` 🤔

akalia25 replied:
I'm seeing an `sg start enterprise` i'll test that as well and see if the results are different!

dadlerj replied:
Makes sense, thanks!

akalia25 commented on client/web/src/tracking/eventLogger.ts:
```diff
@@ -226,29 +228,31 @@ export class EventLogger implements TelemetryService, SharedEventLogger {
-        // Always set to renew expiry and migrate from localStorage
-        cookies.set(FIRST_SOURCE_URL_KEY, redactedURL, this.cookieSettings)
-
-        this.firstSourceURL = firstSourceURL
```
hmm I think the conditional logic I have in place takes care of the removed lines. Instead of always creating the `redactedURL`, we only create it once after determining its not a dotcom source.

dadlerj replied:
I'm actually only referring to the line 235 here, or:

```        this.firstSourceURL = firstSourceURL```

This isn't strictly needed, but it's intended to short-circuit parsing the user's cookies every time an event is logged (which I imagine slightly speeds things up).

dadlerj replied:
I.e., you can see on line 227 (in the old version) / line 229 (in the new version) that the `firstSourceURL` variable is set to the `this.firstSourceURL` value if it exists, and so it only checks the cookies if it hasn't already been set.

akalia25 commented on internal/usagestats/event_handlers.go:
```diff
@@ -321,18 +329,32 @@ func serializeLocalEvents(events []Event) ([]*database.Event, error) {
-//
-// Note that URL redaction also happens in web/src/tracking/util.ts.
+
+// func redactSensitiveInfoFromCloudReferrer (rawURL string) (string, error) {
```
Oops! Yup sorry, this snuck in!

akalia25 commented on internal/usagestats/event_handlers.go:
```diff
@@ -321,18 +329,32 @@ func serializeLocalEvents(events []Event) ([]*database.Event, error) {
+		"info.sourcegraph.com":   true,
+		"signup.sourcegraph.com": true,
+	}
+	if approvedHostsMap[parsedURL.Host] {
```
On that thread, if you look at the second post, I'm following that example, and it seems to be working as expected (based on the tests). Could you elaborate a bit more? 
<img width="734" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/32119652/64403e2c-17a4-4c6e-ba03-42866be08f01">

dadlerj replied:
Ah, thank you for that! That also explains why you used a map with bools. This is fine for me, though it looked a bit weird at first.

akalia25 commented on internal/usagestats/event_handlers.go:
```diff
@@ -343,8 +365,15 @@ func redactSensitiveInfoFromCloudURL(rawURL string) (string, error) {
+		parsedURL.RawPath = "/sign-in/redacted"
+		parsedURL.Path = "/sign-in/redacted"
+	} else if !sourcegraphCloudRegex.MatchString(parsedURL.Host) {
+		// if its not a cloud instance url, return rawURL so that referrers are not redacted
```
ahhh did not realize we moved to `sourcegraphcloud.com` already 🙈 . 

I added that regex specifically for the `referrer` field since it can be especially challenging to redact for cloud instances versus valid traffic when it comes to referrer, as the referrer could potentially be _any_ URL. 

`referrer` is also a significant field for analytics 😬 , so we need to be careful in handling redaction. One scenario I tried keeping in mind is when a user navigates from a cloud instance URL to a sourcegraph.com URL - ensuring the `referrer` field does not contain anything sensitive.

One potential solution for `referrer` is only to track the hostname rather than the full URL, but i'd like to keep referrer if possible 😂 

 Let me know your thoughts!

akalia25 commented on internal/usagestats/event_handlers.go:
```diff
@@ -218,11 +219,18 @@ func serializePublishSourcegraphDotComEvents(events []Event) ([][]byte, error) {
 		if err != nil {
 			return nil, err
 		}
+		saferReferrer, err := redactSensitiveInfoFromCloudURL(referrer)
```
Good call-out will update in the next commit! :)

akalia25 commented on client/web/src/tracking/eventLogger.ts:
```diff
@@ -226,29 +228,31 @@ export class EventLogger implements TelemetryService, SharedEventLogger {
+            cookies.set(FIRST_SOURCE_URL_KEY, firstSourceURL, this.cookieSettings)
+            return firstSourceURL
+        }
+        {
```
Good call-out, thanks!

akalia25 commented:
> Thank you for this! A few small comments, and a few big ones. I want to propose a bit of restructuring though, since I'm afraid that this setup will be a bit brittle (seeing us check if URLs contain specific substrings like "github.com/..." and ".sourcegraph." worries me).
> 
> How does the following sound to simplify it:
> 
> * In the frontend:
>   
>   * If dotcom, send `url` and all marketing URL vars to the backend unredacted.
>   * If cloud, send `url` unredacted, but don't send ANY marketing URL vars (unless you can think of a reason we want them?)
>   * In essence: zero redaction occurs on the frontend and we can centralize it all in the backend
> * In the backend:
>   
>   * If dotcom, no redaction on the `url` var or any of the marketing vars
>   * If cloud, there are no marketing vars. For `url`, we redact using roughly the logic here: https://github.com/sourcegraph/sourcegraph/pull/57384/files#diff-442914ea375af7f025d973a5aec063bcf9814f198b0f057eee99efed7f5862c6R47-R59. The one thing I want to ask is that we expand the list of potential URLs that we pattern match against. The problem is that MANY of our customers don't have `github.com/...` or `gitlab.com/...` prefixes to their code host URLs. They likely host their code internally on an internal domain (e.g. `github.uber.com/...` or just `code.uber.com/...`) so there's no easy way to know what to redact. Let's, at the very least, capture all of Sourcegraph's built-in URL patterns (e.g. notebooks and batch changes and insights and admin pages and user pages and settings pages and etc.) so if we just see the URL as /redacted we know it's most likely the case that it was a custom code view page. Or perhaps instead of trying to list out every possible path that doesn't need to be redacted, we can just always take the first part of the path and redact anything after it—e.g. if the url is `uber.sourcegraphcloud.com/code.uber.com/my-repo-name`, we would take the hostname (uber.sourcegraphcloud.com/), the _first_ part of the path (code.uber.com/), but redact everything after that... Seems like that would work with code views, searches, sign-in, etc.? What do you think?

Thanks for the guidance on this! @dadlerj 
I updated the logic to migrate all redactions to the backend and only set the marketing cookie values when it's dotcom traffic.

In the backend, redaction relies on the `envvar.SourcegraphDotComMode()` to distinguish between dotcom traffic vs cloud instances, and for cloud instances, I expanded the list of potential URLs to pattern match against [here](https://github.com/sourcegraph/sourcegraph/pull/57384/files#diff-4b6f9be139e28324f0b5aefd2bc09b845adf575f8c538c79c8d34de3bdafc80bR377-R383). I used data from our BQ tables to determine what patterns to look for. 

>  take the first part of the path and redact anything after it
>
I believe taking the first part of the path would not work since, within some paths, there is sensitive data in the query parameter (i.e. sign-up, `https://sourcegraph.sourcegraph.com/sign-in?returnTo=%2FtestInfo%2FtestInfo%2`) and the path would include that? We could do some regex to take the first part and then redact as soon as special-characters appear? But I am not sure if that would be hard-coding logic that would go stale in a few years 🤔

dadlerj commented on client/web/src/tracking/util.test.ts:
```diff
@@ -2,59 +2,77 @@ import { getPreviousMonday, redactSensitiveInfoFromAppURL } from './util'
-        it('removes search queries from URLs', () => {
-            expect(redactSensitiveInfoFromAppURL('https://sourcegraph.com/search?q=test+query')).toEqual(
-                'https://sourcegraph.com/search/redacted?q=redacted'
+        it('removes search queries from URLs when originating from cloud instance', () => {
```
Unfortunately all of this great redaction testing is no longer needed on the frontend :). Ideally we'd have comprehensive redaction testing on the backend going forward.

dadlerj replied:
Seems like most (all?) of this file can be removed given redaction is no longer used on the frontend?

dadlerj commented on client/web/src/tracking/util.ts:
```diff
@@ -31,7 +31,16 @@ export function stripURLParameters(url: string, parametersToRemove: string[] = [
 export function redactSensitiveInfoFromAppURL(url: string): string {
```
I believe this whole function can now be deleted. I searched on your branch to see if it was used anywhere else, and it looks like it's just here and in your tests: https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24%40ak/remove-url-redaction+redactSensitiveInfoFromAppURL&patternType=standard&sm=1&groupBy=path

dadlerj commented on internal/usagestats/event_handlers.go:
```diff
@@ -194,31 +194,60 @@ func serializePublishSourcegraphDotComEvents(events []Event) ([][]byte, error) {
 		if event.FirstSourceURL != nil {
 			firstSourceURL = *event.FirstSourceURL
 		}
+		saferFirstSourceURL, err := redactSensitiveInfoFromCloudURL(firstSourceURL)
```
I actually don't think that any of the marketing variables need redaction anymore! Or do they? This change results in us only logging them on dotcom now, if I understand correctly, so they probably don't need any redaction.

dadlerj commented on internal/usagestats/event_handlers.go:
```diff
@@ -194,31 +194,60 @@ func serializePublishSourcegraphDotComEvents(events []Event) ([][]byte, error) {
+		if err != nil {
+			return nil, err
+		}
 		saferUrl, err := redactSensitiveInfoFromCloudURL(event.URL)
```
We would keep this line, though

dadlerj commented on internal/usagestats/event_handlers.go:
```diff
@@ -321,30 +350,46 @@ func serializeLocalEvents(events []Event) ([]*database.Event, error) {
 
-	if parsedURL.Host != "sourcegraph.com" {
+	// check if parsedURL is in approved hosts and if so return rawURL
+	approvedHostsMap := map[string]bool{
```
Is this still needed?

dadlerj commented on internal/usagestats/event_handlers.go:
```diff
@@ -321,30 +350,46 @@ func serializeLocalEvents(events []Event) ([]*database.Event, error) {
-	} else {
-		return rawURL, nil
+	// Define path prefixes to redact
+	redactPaths := []string{
```
Just curious, how did you create this list?

dadlerj replied:
Ah I see above:

> I used data from our BQ tables to determine what patterns to look for.
> 
> take the first part of the path and redact anything after it
> 
> I believe taking the first part of the path would not work since, within some paths, there is sensitive data in the query parameter (i.e. sign-up, https://sourcegraph.sourcegraph.com/sign-in?returnTo=%2FtestInfo%2FtestInfo%2) and the path would include that? We could do some regex to take the first part and then redact as soon as special-characters appear? But I am not sure if that would be hard-coding logic that would go stale in a few years 🤔

Yeah, I don't know of a good method for this. We could just keep the host, first part of the path, and delete _everythign_ after it, including all params. I'd rather we default to _something_ and not be totally blind, I think...

akalia25 commented:
Addressed the comments/suggestions made; thank you @dadlerj!

Highlight of changes:
1. setting cookie value to `this.` value to short circuit reading cookie value
2. removed redundant redaction in backend, since marketing vars are not assigned in front-end unless its dotcom
3. removed `redactSensitiveInfoFromAppURL` from `utils` and the `utils.test` since our redaction function solely lives in the backend
4. Updated `redactSensitiveInfoFromCloudURL` to redact URL paths to only include only the first part of the path ([code change ref](https://github.com/sourcegraph/sourcegraph/pull/57384/files#diff-4b6f9be139e28324f0b5aefd2bc09b845adf575f8c538c79c8d34de3bdafc80bR375-R377)). We apply this right after the URLQueryParams clean-up ([ref](https://github.com/sourcegraph/sourcegraph/pull/57384/files#diff-4b6f9be139e28324f0b5aefd2bc09b845adf575f8c538c79c8d34de3bdafc80bR375-R377)) so that any query parameter outside of our accepted UTM values is redacted. This lets us keep a bit more information but still ensures proper redactions. If we want to remove the redacted query parameters, let me know, and can push that quick fix. 
5. Updated `event_handlers_test.go` to validate we're collecting only the first part of the URL path and that any query parameters outside the accepted ones are being redacted 
 
I only need clarification on `referrer` and if we plan to apply`redactSensitiveInfoFromCloudURL` to this field and if we're aligned on the updates to `redactSensitiveInfoFromCloudURL` :)

dadlerj commented on internal/usagestats/event_handlers.go:
```diff
@@ -321,30 +326,15 @@ func serializeLocalEvents(events []Event) ([]*database.Event, error) {
 
-	if parsedURL.Host != "sourcegraph.com" {
+func redactSensitiveInfoFromCloudURL(rawURL string) (string, error) {
+	if envvar.SourcegraphDotComMode() {
```
```suggestion
	// Because Sourcegraph.com only contains public code, URLs do not contain sensitive information.
	// Redaction is only used for URLs from cloud and self-hosted instance telemetry.
	if envvar.SourcegraphDotComMode() {
```

dadlerj commented on internal/usagestats/event_handlers.go:
```diff
@@ -364,12 +354,17 @@ func redactSensitiveInfoFromCloudURL(rawURL string) (string, error) {
 	}
 
+	// Retain only the first part of the URL's path segment for security.
+	parsedURL.Path = strings.Split(parsedURL.Path, "/")[1]
```
Are you sure you don't want to take the 0th element in the array instead of the 1st here? If the path doesn't contain "/" at all, then the `Split` function would return an array with only 1 element and accessing the `[1]` element (instead of `[0]`) would throw an error I think.

Edit: 

I see below that this passes your tests... I'm just a bit confused about why this works. Is it because `parsedURL.Path` always _starts_ with a `/` and so the 0th element is just an empty string?

If so, I might add a check to ensure that the result contains 2 elements before doing this, _just in case_ something changes someday... I wouldn't want us to cause a customer's server to break down because we reference a non-existent array element.

dadlerj replied:
Separately, if we actually end up deleting part of the path here (so if there are extra elements in this array that we don't include), then can you add `/redacted` to the end of the resulting path string?

akalia25 replied:
Definitely do not want to break customer's server with array index issues! Added a check to make sure the length is sufficient before redacting and if we are redacting more than one part of the path, then the suffix `redacted` is added. (i.e. `soucegraph.com/first/second/third`  --> `sourcegraph.com/first/redacted`. I did not see value in populating a `/redacted` for each additional part of the path, but if there's value I can update accordingly

dadlerj commented on internal/usagestats/event_handlers_test.go:
```diff
@@ -14,19 +14,39 @@ func TestRedactSensitiveInfoFromCloudURL(t *testing.T) {
-			name: "path redacted",
-			url:  "https://sourcegraph.com/github.com/test/test",
-			want: "https://sourcegraph.com/github.com/redacted",
+			name: "URL path parameters are redacted",
```
```suggestion
			name: "URL path parameters are redacted (when `SourcegraphDotComMode` is false)",
```

dadlerj replied:
This test is a bit confusing since sourcegraph.com URLs usually _wouldn't_ be redacted. I added this explanation to the test name.

dadlerj reviewed (commented):
Thank you! I think we're super close. Just left a few more comments.

akalia25 commented on client/web/src/tracking/eventLogger.ts:
```diff
@@ -225,30 +227,26 @@ export class EventLogger implements TelemetryService, SharedEventLogger {
 
     public getFirstSourceURL(): string {
```
Great point! I re-ordered it as per your suggestion :)

dadlerj reviewed (approved):
Thanks! I think this looks good to me pending a response to the open question in Slack
